inference-engine/thirdparty/clDNN/src/gpu/kernels_cache.cpp

```diff
@@ -140,23 +242,30 @@ kernels_cache::sorted_code kernels_cache::get_program_source(const kernels_code&
             current_bucket.options = options;
         }
 
-        if ((current_bucket.kernels_counter % MAX_KERNELS_PER_PROGRAM) == 0) {
+        // Create new kernels bucket when the limit is reached
+        if ((current_bucket.kernels_counter % get_max_kernels_per_batch()) == 0) {
             current_bucket.source.push_back({});
         }
 
         current_bucket.entry_point_to_id[entry_point] = code.id;
+        assert(org_source_code.size() == 1);
 
-        source_code new_source_code = org_source_code;
+        current_bucket.source.back().push_back(std::move(org_source_code.front()));
 
-        if (batch_compilation) {
-            new_source_code.push_back(get_undef_jit(org_source_code));
-        }
+        current_bucket.kernels_counter++;
+    }
 
-        for (auto& s : new_source_code) {
-            current_bucket.source.back().push_back(std::move(s));
+    // Compute hash value for each bucket
+    // Hash calculation might require additional optimizations, but currently execution time of this part is much smaller than loading
+    // of the precompiled binaries or get_undef_jit calls
+    for (auto& c : scode) {
+        program_code& code = c.second;
+        for (size_t i = 0; i < code.source.size(); i++) {
+            std::string full_code;
+            for (auto& ss : code.source[i])
+                full_code += ss;
+            code.hash_values.push_back(std::hash<std::string>()(full_code));
```

Comment:
BTW, hash is a source code? Maybe we needs some notes about has compute part, e.g. how it's computed

Reply:
added a comment that explains what exactly is used for hash computation. Do you think it should be duplicated somewhere in the docs or we don't need such kind of implementation details in the docs?

Reply:
I think inline comments in code are enough.